src/main/java/org/folio/search/service/consortium/ConsortiumInstanceSearchService.java

```diff
@@ -67,34 +89,105 @@ public ConsortiumItem getConsortiumItem(String id, CqlSearchRequest<Instance> se
+    var targetField = INSTANCE_HOLDING_FIELD_NAME + ".id";
+
+    var searchRecords = getConsortiumBatchResults(tenant, ids, targetField, this::mapToConsortiumHolding);
+
+    if (searchRecords.isEmpty()) {
+      return new ConsortiumHoldingCollection();
+    }
+
     return new ConsortiumHoldingCollection()
-      .holdings(consortiumHoldings)
-      .totalRecords(consortiumHoldings.size());
+      .holdings(searchRecords.stream().map(SearchResult::getRecords).flatMap(List::stream).toList())
+      .totalRecords(searchRecords.iterator().next().getTotalRecords());
   }
 
-  public ConsortiumItemCollection fetchConsortiumBatchItems(CqlSearchRequest<Instance> searchRequest,
-                                                            Set<String> ids) {
-    var result = searchService.search(searchRequest);
-    var consortiumItems = result.getRecords().stream()
-      .flatMap(instance ->
-        instance.getItems().stream()
-          .filter(item -> ids.contains(item.getId()))
-          .map(item -> ConsortiumItemMapper.toConsortiumItem(instance.getId(), item))
-      )
-      .toList();
+  public ConsortiumItemCollection fetchConsortiumBatchItems(String tenant, Set<UUID> itemIds) {
+    validateIdsCount(itemIds.size());
+
+    var ids = itemIds.stream().map(UUID::toString).collect(Collectors.toSet());
+    var targetField = INSTANCE_ITEM_FIELD_NAME + ".id";
+
+    var searchRecords = getConsortiumBatchResults(tenant, ids, targetField, this::mapToConsortiumItem);
+
+    if (searchRecords.isEmpty()) {
+      return new ConsortiumItemCollection();
+    }
 
     return new ConsortiumItemCollection()
-      .items(consortiumItems)
-      .totalRecords(consortiumItems.size());
+      .items(searchRecords.stream().map(SearchResult::getRecords).flatMap(List::stream).toList())
+      .totalRecords(searchRecords.iterator().next().getTotalRecords());
+  }
+
+  private <T> List<SearchResult<T>> getConsortiumBatchResults(String tenant, Set<String> ids, String targetField,
+                                                              BiFunction<Instance, Set<String>, List<T>> recordMapper) {
+    var request = CqlSearchRequest.of(Instance.class, tenant, "", 0, 0, true, false, true);
+    var termsQuery = termsQuery(targetField, ids);
+
+    if (ids.size() < DEFAULT_MAX_SEARCH_RESULT_WINDOW) {
+      var searchSourceBuilder = queryBuilder(termsQuery, ids.size());
+      var response = searchRepository.search(request, searchSourceBuilder);
+      var searchResult = documentConverter.convertToSearchResult(response, request.getResourceClass(),
+        (hits, item) -> recordMapper.apply(item, ids));
+      var records = searchResult.getRecords().stream()
+        .flatMap(List::stream)
+        .toList();
+      return List.of(SearchResult.of(records.size(), records));
+    }
+
+    var searchSourceBuilder = queryBuilder(termsQuery, properties.getSearchConsortiumRecordsPageSize())
```

Comment:
Why are we querying 5_000 (default) here, while above we can query more while it's under 10_000, even 9_000?